frontend/src/metabase/admin/datamodel/components/database/MetadataTableList/MetadataTableList.tsx

```diff
@@ -0,0 +1,166 @@
+import React, { ChangeEventHandler, useState } from "react";
+import { connect } from "react-redux";
+import { PLUGIN_ADVANCED_PERMISSIONS } from "metabase/plugins";
+
+import Tables from "metabase/entities/tables";
+
+import Icon from "metabase/components/Icon";
+
+import { t, ngettext, msgid } from "ttag";
+
+import _ from "underscore";
+
+import { regexpEscape } from "metabase/lib/string";
+import { ToggleVisibilityButton } from "./ToggleVisibilityButton";
+import { TableRow } from "./TableRow";
+import { Table } from "metabase-types/types/Table";
+
+interface MetadataTableListProps {
+  onBack?: () => void;
+  schema?: string;
+  setVisibilityForTables: (tables: Table[]) => void;
+  tables: Table[];
+  tableId: number;
+  selectTable: (table: Table) => void;
+}
+
+const MetadataTableList = ({
+  onBack,
+  schema,
+  tables,
+  tableId,
+  selectTable,
+  setVisibilityForTables,
+}: MetadataTableListProps) => {
+  const [search, setSearch] = useState<{ text: string; regex: RegExp | null }>({
+    text: "",
+    regex: null,
+  });
+
+  const handleSearchChange: ChangeEventHandler<HTMLInputElement> = event => {
+    setSearch({
+      text: event.target.value,
+      regex: event.target.value
+        ? new RegExp(regexpEscape(event.target.value), "i")
```

Comment:
It seems that `regex` is not state, and should be computed and memoized from `text`